apps/namadillo/src/atoms/balance/atoms.ts

```diff
@@ -148,75 +153,84 @@ export const tokenPricesAtom = atomWithQuery((get) => {
+  const chainParameters = get(chainParametersAtom);
 
   return {
     queryKey: [
-      "shielded-tokens",
-      shieldedBalanceQuery.data,
-      assetByAddressQuery.data,
-      tokenPricesQuery.data,
+      "namada-shielded-assets",
+      shieldedBalances.data,
+      tokenAddresses.data,
+      chainParameters.data!.chainId,
     ],
     ...queryDependentFn(
       async () =>
-        shieldedBalanceQuery.data?.map(({ address, amount }) =>
-          formatTokenBalance(
-            address,
-            amount,
-            assetByAddressQuery.data,
-            tokenPricesQuery.data
-          )
-        ) ?? [],
-      [shieldedBalanceQuery, tokenPricesQuery, assetByAddressQuery]
+        await mapNamadaAddressesToAssets(
+          shieldedBalances.data!,
+          tokenAddresses.data!,
+          chainParameters.data!.chainId
+        ),
+      [shieldedBalances, tokenAddresses, chainParameters]
     ),
   };
 });
 
-export const transparentTokensAtom = atomWithQuery<TokenBalance[]>((get) => {
-  const transparentBalanceQuery = get(transparentBalanceAtom);
-  const assetByAddressQuery = get(assetByAddressAtom);
-  const tokenPricesQuery = get(tokenPricesAtom);
+export const namadaTransparentAssetsAtom = atomWithQuery((get) => {
+  const transparentBalances = get(transparentBalanceAtom);
+  const tokenAddresses = get(tokenAddressesAtom);
+  const chainParameters = get(chainParametersAtom);
 
   return {
     queryKey: [
-      "transparent-tokens",
-      transparentBalanceQuery.data,
-      assetByAddressQuery.data,
-      tokenPricesQuery.data,
+      "namada-transparent-assets",
+      transparentBalances.data,
+      tokenAddresses.data,
+      chainParameters.data!.chainId,
     ],
     ...queryDependentFn(
       async () =>
-        transparentBalanceQuery.data?.map(({ address, amount }) =>
-          formatTokenBalance(
-            address,
-            amount,
-            assetByAddressQuery.data,
-            tokenPricesQuery.data
-          )
-        ) ?? [],
-      [transparentBalanceQuery, tokenPricesQuery, assetByAddressQuery]
+        await mapNamadaAddressesToAssets(
+          transparentBalances.data!,
+          tokenAddresses.data!,
+          chainParameters.data!.chainId
+        ),
+      [transparentBalances, tokenAddresses, chainParameters]
     ),
   };
 });
 
-const formatTokenBalance = (
-  address: string,
-  amount: BigNumber,
-  assetsByAddress?: Record<string, Asset>,
-  tokenPrices?: Record<string, BigNumber>
-): TokenBalance => {
-  const asset = assetsByAddress?.[address] ?? unknownAsset;
-  const balance = toDisplayAmount(asset, amount);
+export const shieldedTokensAtom = atomWithQuery<TokenBalance[]>((get) => {
+  const shieldedAssets = get(namadaShieldedAssetsAtom);
+  const tokenPrices = get(tokenPricesAtom);
+
+  return {
+    queryKey: ["shielded-tokens", shieldedAssets.data, tokenPrices.data],
+    ...queryDependentFn(
+      async () =>
+        await mapNamadaAssetsToTokenBalances(
```

Comment:
This is not an async operation, it's just a sync transformation. Then maybe we should have if as a helper function than an atom, like we do with `getTotalDollar`. What do you think?

Reply:
Nice catch! I was thinking if we do make it a helper function, we'll need to copy the code to get `shieldedAssets` and `tokenPrices` everywhere it is used, so since it's two args it might be a bit less nice than `getTotalDollar`.